Filtering observations - Time out exception

```
{
  dataset_births {
    observations(dimensions: {gender: MALE, 
                              reference_period:{starts_after:  "2000-01-01T00:00:00Z"
                                                ends_before: "2001-01-01T00:00:00Z"}}) {
      page(first: 20) {
        observations {
          reference_area {
            label
            uri
          }
          reference_period {
            start
            end
            label
          }
        }
      }
    }
  }
}
```

@lkitching I am trying to solve this issue but there is not much documentation on the code to understand how bindings are treated at the code.

I would like to create SPARQL statements like this:
```
?obs <http://purl.org/linked-data/cube#measureType> ?mp .
?obs ?mp ?mv
```

I changed the code for the measure type at `types.clj`
```
(defrecord MeasureType [uri label order is-numeric?]
  ...
  SparqlResultProjector
  (apply-projection [_this model selections]
    (let [dim-key (keyword (str "mv" order))
           dim-value (keyword (str "mp" order))]
       (let [model (qm/add-binding model [[dim-value (URI. "http://purl.org/linked-data/cube#measureType")]] ::qm/var)
             model (qm/add-binding model [[dim-key dim-value]] ::qm/var)]
         model)))    
```

But I get an exception when running the code. Any idea what causes the exception and how can be solved?

```
clojure.lang.ExceptionInfo: java.lang.IllegalArgumentException in Interceptor :com.walmartlabs.lacinia.pedestal/query-executor - No implementation of method: :query-format of protocol: #'graphql-qb.query-model/QueryItem found for class: clojure.lang.Keyword
	at clojure.core$ex_info.invokeStatic(core.clj:4739) ~[clojure-1.9.0.jar:?]
	at clojure.core$ex_info.invoke(core.clj:4739) ~[clojure-1.9.0.jar:?]
	at io.pedestal.interceptor.chain$throwable__GT_ex_info.invokeStatic(chain.clj:35) ~[?:?]
	at io.pedestal.interceptor.chain$throwable__GT_ex_info.invoke(chain.clj:32) ~[?:?]
	at io.pedestal.interceptor.chain$try_f.invokeStatic(chain.clj:57) ~[?:?]
	at io.pedestal.interceptor.chain$try_f.invoke(chain.clj:44) ~[?:?]
	at io.pedestal.interceptor.chain$process_all_with_binding.invokeStatic(chain.clj:171) ~[?:?]
	at io.pedestal.interceptor.chain$process_all_with_binding.invoke(chain.clj:146) ~[?:?]
	at io.pedestal.interceptor.chain$process_all$fn__17645.invoke(chain.clj:188) ~[?:?]
	at clojure.lang.AFn.applyToHelper(AFn.java:152) ~[clojure-1.9.0.jar:?]
	at clojure.lang.AFn.applyTo(AFn.java:144) ~[clojure-1.9.0.jar:?]
	at clojure.core$apply.invokeStatic(core.clj:657) ~[clojure-1.9.0.jar:?]
	at clojure.core$with_bindings_STAR_.invokeStatic(core.clj:1965) ~[clojure-1.9.0.jar:?]
	at clojure.core$with_bindings_STAR_.doInvoke(core.clj:1965) ~[clojure-1.9.0.jar:?]
	at clojure.lang.RestFn.invoke(RestFn.java:425) ~[clojure-1.9.0.jar:?]
	at io.pedestal.interceptor.chain$process_all.invokeStatic(chain.clj:186) ~[?:?]
	at io.pedestal.interceptor.chain$process_all.invoke(chain.clj:182) ~[?:?]
	at io.pedestal.interceptor.chain$enter_all.invokeStatic(chain.clj:235) ~[?:?]
	at io.pedestal.interceptor.chain$enter_all.invoke(chain.clj:229) ~[?:?]
	at io.pedestal.interceptor.chain$execute.invokeStatic(chain.clj:379) [?:?]
	at io.pedestal.interceptor.chain$execute.invoke(chain.clj:352) [?:?]
	at io.pedestal.interceptor.chain$execute.invokeStatic(chain.clj:389) [?:?]
	at io.pedestal.interceptor.chain$execute.invoke(chain.clj:352) [?:?]
	at io.pedestal.http.impl.servlet_interceptor$interceptor_service_fn$fn__21889.invoke(servlet_interceptor.clj:350) [?:?]
	at io.pedestal.http.servlet.FnServlet.service(servlet.clj:28) [?:?]
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:838) [jetty-servlet-9.4.0.v20161208.jar:9.4.0.v20161208]
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:543) [jetty-servlet-9.4.0.v20161208.jar:9.4.0.v20161208]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:188) [jetty-server-9.4.0.v20161208.jar:9.4.0.v20161208]
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1228) [jetty-server-9.4.0.v20161208.jar:9.4.0.v20161208]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:168) [jetty-server-9.4.0.v20161208.jar:9.4.0.v20161208]
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:481) [jetty-servlet-9.4.0.v20161208.jar:9.4.0.v20161208]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:166) [jetty-server-9.4.0.v20161208.jar:9.4.0.v20161208]
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1130) [jetty-server-9.4.0.v20161208.jar:9.4.0.v20161208]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141) [jetty-server-9.4.0.v20161208.jar:9.4.0.v20161208]
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132) [jetty-server-9.4.0.v20161208.jar:9.4.0.v20161208]
	at org.eclipse.jetty.server.Server.handle(Server.java:564) [jetty-server-9.4.0.v20161208.jar:9.4.0.v20161208]
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:318) [jetty-server-9.4.0.v20161208.jar:9.4.0.v20161208]
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:251) [jetty-server-9.4.0.v20161208.jar:9.4.0.v20161208]
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:279) [jetty-io-9.4.0.v20161208.jar:9.4.0.v20161208]
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:112) [jetty-io-9.4.0.v20161208.jar:9.4.0.v20161208]
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:124) [jetty-io-9.4.0.v20161208.jar:9.4.0.v20161208]
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:672) [jetty-util-9.4.0.v20161208.jar:9.4.0.v20161208]
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:590) [jetty-util-9.4.0.v20161208.jar:9.4.0.v20161208]
	at java.lang.Thread.run(Thread.java:745) [?:1.8.0_65]
Caused by: java.lang.IllegalArgumentException: No implementation of method: :query-format of protocol: #'graphql-qb.query-model/QueryItem found for class: clojure.lang.Keyword
	at clojure.core$_cache_protocol_fn.invokeStatic(core_deftype.clj:583) ~[clojure-1.9.0.jar:?]
	at clojure.core$_cache_protocol_fn.invoke(core_deftype.clj:575) ~[clojure-1.9.0.jar:?]
	at graphql_qb.query_model$eval5457$fn__5458$G__5448__5463.invoke(query_model.clj:11) ~[?:?]
	at graphql_qb.query_model$format_query_triple_pattern.invokeStatic(query_model.clj:148) ~[?:?]
	at graphql_qb.query_model$format_query_triple_pattern.invoke(query_model.clj:147) ~[?:?]
	at clojure.core$map$fn__5587.invoke(core.clj:2747) ~[clojure-1.9.0.jar:?]
	at clojure.lang.LazySeq.sval(LazySeq.java:40) ~[clojure-1.9.0.jar:?]
	at clojure.lang.LazySeq.seq(LazySeq.java:49) ~[clojure-1.9.0.jar:?]
	at clojure.lang.Cons.next(Cons.java:39) ~[clojure-1.9.0.jar:?]
	at clojure.lang.RT.next(RT.java:706) ~[clojure-1.9.0.jar:?]
	at clojure.core$next__5108.invokeStatic(core.clj:64) ~[clojure-1.9.0.jar:?]
	at clojure.string$join.invokeStatic(string.clj:191) ~[clojure-1.9.0.jar:?]
	at clojure.string$join.invoke(string.clj:180) ~[clojure-1.9.0.jar:?]
	at graphql_qb.query_model$get_query.invokeStatic(query_model.clj:177) ~[?:?]
	at graphql_qb.query_model$get_query.invoke(query_model.clj:167) ~[?:?]
	at graphql_qb.queries$get_observation_query.invokeStatic(queries.clj:33) ~[?:?]
	at graphql_qb.queries$get_observation_query.invoke(queries.clj:31) ~[?:?]
	at graphql_qb.queries$get_observation_page_query.invokeStatic(queries.clj:37) ~[?:?]
	at graphql_qb.queries$get_observation_page_query.invoke(queries.clj:35) ~[?:?]
	at graphql_qb.resolvers$inner_resolve_observations_page.invokeStatic(resolvers.clj:102) ~[?:?]
	at graphql_qb.resolvers$inner_resolve_observations_page.invoke(resolvers.clj:96) ~[?:?]
	at graphql_qb.resolvers$wrap_pagination_resolver$fn__6498.invoke(resolvers.clj:90) ~[?:?]
	at graphql_qb.schema$wrap_observations_mapping$fn__6653.invoke(schema.clj:59) ~[?:?]
	at com.walmartlabs.lacinia.schema$wrap_resolver_to_ensure_resolver_result$fn__3656.invoke(schema.clj:519) ~[?:?]
	at com.walmartlabs.lacinia.executor$invoke_resolver_for_field.invokeStatic(executor.clj:114) ~[?:?]
	at com.walmartlabs.lacinia.executor$invoke_resolver_for_field.invoke(executor.clj:86) ~[?:?]
	at com.walmartlabs.lacinia.executor$resolve_and_select.invokeStatic(executor.clj:421) ~[?:?]
	at com.walmartlabs.lacinia.executor$resolve_and_select.invoke(executor.clj:319) ~[?:?]
	at com.walmartlabs.lacinia.executor$eval5149$fn__5150.invoke(executor.clj:203) ~[?:?]
	at clojure.lang.MultiFn.invoke(MultiFn.java:233) ~[clojure-1.9.0.jar:?]
	at com.walmartlabs.lacinia.executor$maybe_apply_selection.invokeStatic(executor.clj:257) ~[?:?]
	at com.walmartlabs.lacinia.executor$maybe_apply_selection.invoke(executor.clj:253) ~[?:?]
	at com.walmartlabs.lacinia.executor$execute_nested_selections$fn__5174.invoke(executor.clj:281) ~[?:?]
	at clojure.core$keep$fn__8200.invoke(core.clj:7261) ~[clojure-1.9.0.jar:?]
	at clojure.lang.LazySeq.sval(LazySeq.java:40) ~[clojure-1.9.0.jar:?]
	at clojure.lang.LazySeq.seq(LazySeq.java:49) ~[clojure-1.9.0.jar:?]
	at clojure.lang.RT.seq(RT.java:528) ~[clojure-1.9.0.jar:?]
	at clojure.core$seq__5124.invokeStatic(core.clj:137) ~[clojure-1.9.0.jar:?]
	at clojure.core.protocols$seq_reduce.invokeStatic(protocols.clj:24) ~[clojure-1.9.0.jar:?]
	at clojure.core.protocols$fn__7835.invokeStatic(protocols.clj:75) ~[clojure-1.9.0.jar:?]
	at clojure.core.protocols$fn__7835.invoke(protocols.clj:75) ~[clojure-1.9.0.jar:?]
	at clojure.core.protocols$fn__7781$G__7776__7794.invoke(protocols.clj:13) ~[clojure-1.9.0.jar:?]
	at clojure.core$reduce.invokeStatic(core.clj:6748) ~[clojure-1.9.0.jar:?]
	at clojure.core$reduce.invoke(core.clj:6730) ~[clojure-1.9.0.jar:?]
	at com.walmartlabs.lacinia.executor$execute_nested_selections.invokeStatic(executor.clj:282) ~[?:?]
	at com.walmartlabs.lacinia.executor$execute_nested_selections.invoke(executor.clj:273) ~[?:?]
	at com.walmartlabs.lacinia.executor$resolve_and_select$selector_callback__5219.invoke(executor.clj:344) ~[?:?]
	at com.walmartlabs.lacinia.schema$floor_selector.invokeStatic(schema.clj:528) ~[?:?]
	at com.walmartlabs.lacinia.schema$floor_selector.invoke(schema.clj:525) ~[?:?]
	at com.walmartlabs.lacinia.schema$create_root_selector$select_unwrap_tagged_type__3668.invoke(schema.clj:614) ~[?:?]
	at com.walmartlabs.lacinia.schema$create_root_selector$select_apply_static_type__3670.invoke(schema.clj:619) ~[?:?]
	at com.walmartlabs.lacinia.schema$create_root_selector$select_require_single_value__3673.invoke(schema.clj:627) ~[?:?]
	at com.walmartlabs.lacinia.executor$resolve_and_select$process_resolved_value__5226.invoke(executor.clj:385) ~[?:?]
	at com.walmartlabs.lacinia.executor$resolve_and_select$receive_resolved_value_from_field__5228.invoke(executor.clj:423) ~[?:?]
	at com.walmartlabs.lacinia.resolve.ResolverResultImpl.on_deliver_BANG_(resolve.clj:94) ~[?:?]
	at com.walmartlabs.lacinia.executor$resolve_and_select.invokeStatic(executor.clj:421) ~[?:?]
	at com.walmartlabs.lacinia.executor$resolve_and_select.invoke(executor.clj:319) ~[?:?]
	at com.walmartlabs.lacinia.executor$eval5149$fn__5150.invoke(executor.clj:203) ~[?:?]
	at clojure.lang.MultiFn.invoke(MultiFn.java:233) ~[clojure-1.9.0.jar:?]
	at com.walmartlabs.lacinia.executor$maybe_apply_selection.invokeStatic(executor.clj:257) ~[?:?]
	at com.walmartlabs.lacinia.executor$maybe_apply_selection.invoke(executor.clj:253) ~[?:?]
	at com.walmartlabs.lacinia.executor$execute_nested_selections$fn__5174.invoke(executor.clj:281) ~[?:?]
	at clojure.core$keep$fn__8200.invoke(core.clj:7257) ~[clojure-1.9.0.jar:?]
	at clojure.lang.LazySeq.sval(LazySeq.java:40) ~[clojure-1.9.0.jar:?]
	at clojure.lang.LazySeq.seq(LazySeq.java:49) ~[clojure-1.9.0.jar:?]
	at clojure.lang.RT.seq(RT.java:528) ~[clojure-1.9.0.jar:?]
	at clojure.core$seq__5124.invokeStatic(core.clj:137) ~[clojure-1.9.0.jar:?]
	at clojure.core.protocols$seq_reduce.invokeStatic(protocols.clj:24) ~[clojure-1.9.0.jar:?]
	at clojure.core.protocols$fn__7835.invokeStatic(protocols.clj:75) ~[clojure-1.9.0.jar:?]
	at clojure.core.protocols$fn__7835.invoke(protocols.clj:75) ~[clojure-1.9.0.jar:?]
	at clojure.core.protocols$fn__7781$G__7776__7794.invoke(protocols.clj:13) ~[clojure-1.9.0.jar:?]
	at clojure.core$reduce.invokeStatic(core.clj:6748) ~[clojure-1.9.0.jar:?]
	at clojure.core$reduce.invoke(core.clj:6730) ~[clojure-1.9.0.jar:?]
	at com.walmartlabs.lacinia.executor$execute_nested_selections.invokeStatic(executor.clj:282) ~[?:?]
	at com.walmartlabs.lacinia.executor$execute_nested_selections.invoke(executor.clj:273) ~[?:?]
	at com.walmartlabs.lacinia.executor$resolve_and_select$selector_callback__5219.invoke(executor.clj:344) ~[?:?]
	at com.walmartlabs.lacinia.schema$floor_selector.invokeStatic(schema.clj:528) ~[?:?]
	at com.walmartlabs.lacinia.schema$floor_selector.invoke(schema.clj:525) ~[?:?]
	at com.walmartlabs.lacinia.schema$create_root_selector$select_unwrap_tagged_type__3668.invoke(schema.clj:614) ~[?:?]
	at com.walmartlabs.lacinia.schema$create_root_selector$select_apply_static_type__3670.invoke(schema.clj:619) ~[?:?]
	at com.walmartlabs.lacinia.schema$create_root_selector$select_require_single_value__3673.invoke(schema.clj:627) ~[?:?]
	at com.walmartlabs.lacinia.executor$resolve_and_select$process_resolved_value__5226.invoke(executor.clj:385) ~[?:?]
	at com.walmartlabs.lacinia.executor$resolve_and_select$receive_resolved_value_from_field__5228.invoke(executor.clj:423) ~[?:?]
	at com.walmartlabs.lacinia.resolve.ResolverResultImpl.on_deliver_BANG_(resolve.clj:94) ~[?:?]
	at com.walmartlabs.lacinia.executor$resolve_and_select.invokeStatic(executor.clj:421) ~[?:?]
	at com.walmartlabs.lacinia.executor$resolve_and_select.invoke(executor.clj:319) ~[?:?]
	at com.walmartlabs.lacinia.executor$eval5149$fn__5150.invoke(executor.clj:203) ~[?:?]
	at clojure.lang.MultiFn.invoke(MultiFn.java:233) ~[clojure-1.9.0.jar:?]
	at com.walmartlabs.lacinia.executor$maybe_apply_selection.invokeStatic(executor.clj:257) ~[?:?]
	at com.walmartlabs.lacinia.executor$maybe_apply_selection.invoke(executor.clj:253) ~[?:?]
	at com.walmartlabs.lacinia.executor$execute_nested_selections$fn__5174.invoke(executor.clj:281) ~[?:?]
	at clojure.core$keep$fn__8200.invoke(core.clj:7257) ~[clojure-1.9.0.jar:?]
	at clojure.lang.LazySeq.sval(LazySeq.java:40) ~[clojure-1.9.0.jar:?]
	at clojure.lang.LazySeq.seq(LazySeq.java:49) ~[clojure-1.9.0.jar:?]
	at clojure.lang.RT.seq(RT.java:528) ~[clojure-1.9.0.jar:?]
	at clojure.core$seq__5124.invokeStatic(core.clj:137) ~[clojure-1.9.0.jar:?]
	at clojure.core.protocols$seq_reduce.invokeStatic(protocols.clj:24) ~[clojure-1.9.0.jar:?]
	at clojure.core.protocols$fn__7835.invokeStatic(protocols.clj:75) ~[clojure-1.9.0.jar:?]
	at clojure.core.protocols$fn__7835.invoke(protocols.clj:75) ~[clojure-1.9.0.jar:?]
	at clojure.core.protocols$fn__7781$G__7776__7794.invoke(protocols.clj:13) ~[clojure-1.9.0.jar:?]
	at clojure.core$reduce.invokeStatic(core.clj:6748) ~[clojure-1.9.0.jar:?]
	at clojure.core$reduce.invoke(core.clj:6730) ~[clojure-1.9.0.jar:?]
	at com.walmartlabs.lacinia.executor$execute_nested_selections.invokeStatic(executor.clj:282) ~[?:?]
	at com.walmartlabs.lacinia.executor$execute_nested_selections.invoke(executor.clj:273) ~[?:?]
	at com.walmartlabs.lacinia.executor$execute_query.invokeStatic(executor.clj:454) ~[?:?]
	at com.walmartlabs.lacinia.executor$execute_query.invoke(executor.clj:428) ~[?:?]
	at com.walmartlabs.lacinia.pedestal$execute_query.invokeStatic(pedestal.clj:239) ~[?:?]
	at com.walmartlabs.lacinia.pedestal$execute_query.invoke(pedestal.clj:234) ~[?:?]
	at com.walmartlabs.lacinia.pedestal$fn__23320.invokeStatic(pedestal.clj:254) ~[?:?]
	at com.walmartlabs.lacinia.pedestal$fn__23320.invoke(pedestal.clj:253) ~[?:?]
	at io.pedestal.interceptor.chain$try_f.invokeStatic(chain.clj:54) ~[?:?]
	... 39 more
```

--- a/src/graphql_qb/queries.clj
+++ b/src/graphql_qb/queries.clj
@@ -3,13 +3,17 @@
             [graphql-qb.types :as types]
             [grafter.rdf.sparql :as sp]
             [graphql-qb.vocabulary :refer :all]
-            [graphql-qb.query-model :as qm]))
+            [graphql-qb.query-model :as qm])
+  (:import  [java.net URI]))
 
 (defn get-observation-filter-model [dim-filter]
-  (reduce (fn [m [dim value]]
-            (types/apply-filter dim m value))
-          qm/empty-model
-          dim-filter))
+   (let [m (-> qm/empty-model
+                    (qm/add-binding [[:mp (URI. "http://purl.org/linked-data/cube#measureType")]] ::qm/var)
+                    (qm/add-binding [[:mv (qm/->QueryVar "mp")]] ::qm/var))]
+       (reduce (fn [m [dim value]]
+                       (types/apply-filter dim m value))
+                     m
+                     dim-filter)))
 
 (defn apply-model-projections [filter-model dataset observation-selections]
   (reduce (fn [m dm]

--- a/src/graphql_qb/schema.clj
+++ b/src/graphql_qb/schema.clj
@@ -57,7 +57,8 @@
 (defn wrap-observations-mapping [inner-resolver dataset dataset-enum-mappings]
   (fn [context args observations-field]
     (let [result (inner-resolver context args observations-field)
-          projection (merge {:uri :obs} (types/dataset-result-projection dataset))
+          updated-result (first (::resolvers/observation-results result))
+          projection (merge {:uri :obs} (types/dataset-result-projection dataset updated-result))
           result-mapping (mapping/get-dataset-observations-result-mapping dataset dataset-enum-mappings)
           mapped-result (mapv (fn [obs-bindings]
                                 (let [sparql-result (types/project-result projection obs-bindings)]

--- a/src/graphql_qb/types.clj
+++ b/src/graphql_qb/types.clj
@@ -79,7 +79,7 @@
 (defprotocol SparqlResultProjector
   (apply-projection [this model selections])
   (project-result [this sparql-binding])
-  (get-result-projection [this]))
+  (get-result-projection [this sparql-binding]))
 
 (extend-protocol SparqlResultProjector
   Keyword
@@ -120,7 +120,7 @@
     (let [dim-key (keyword (str "dim" order))]
       (if (is-ref-area-type? type)
         (-> model
-            (qm/add-binding [[dim-key uri] [:label rdfs:label]] ::qm/var :optional? true)
+            (qm/add-binding [[dim-key uri] [:label rdfs:label]] ::qm/var)
             (qm/add-order-by {direction [dim-key :label]}))
         ;;NOTE: binding should have already been added
         (qm/add-order-by model {direction [dim-key]}))))
@@ -142,7 +142,7 @@
         (is-ref-period-type? type)
         (let [model (qm/add-binding model [[dim-key uri]] ::qm/var)
               model (if (contains? field-selections :label)
-                      (qm/add-binding model [[dim-key uri] [:label rdfs:label]] ::qm/var :optional? true)
+                      (qm/add-binding model [[dim-key uri] [:label rdfs:label]] ::qm/var)
                       model)
               model (if (contains? field-selections :start)
                       (qm/add-binding model [[dim-key uri] [:begin time:hasBeginning] [:time time:inXSDDateTime]] ::qm/var)
@@ -154,7 +154,7 @@
         (is-ref-area-type? type)
         (let [label-selected? (contains? field-selections :label)]
           (if label-selected?
-            (qm/add-binding model [[dim-key uri] [:label rdfs:label]] ::qm/var :optional? true)
+            (qm/add-binding model [[dim-key uri] [:label rdfs:label]] ::qm/var)
             model))
 
         :else model)))
@@ -174,7 +174,7 @@
 
         :else (get bindings dim-key))))
 
-  (get-result-projection [_this]
+  (get-result-projection [_this bindings]
     (let [dim-key (keyword (str "dim" order))]
       (cond
         (is-ref-period-type? type)
@@ -192,19 +192,22 @@
 (defrecord MeasureType [uri label order is-numeric?]
   SparqlQueryable
   (apply-order-by [this model direction]
-    (qm/add-order-by model {direction [(keyword (str "mv" order))]}))
+    (qm/add-order-by model {direction [(keyword (str "mv"))]}))
 
   SparqlResultProjector
   (apply-projection [_this model selections]
-    (let [dim-key (keyword (str "mv" order))]
-      (qm/add-binding model [[dim-key uri]] ::qm/var :optional? true)))
+    model)
 
   (project-result [_this binding]
-    (get binding (keyword (str "mv" order))))
+    (if (= uri (get binding :mp))
+             (get binding :mv)
+             nil))
 
-  (get-result-projection [_this]
-    (let [dim-key (keyword (str "mv" order))]
-      (->PathProjection [[dim-key uri]] true identity))))
+  (get-result-projection [_this bindings]
+    (if (= uri (get bindings :mp))
+      (let [dim-key (keyword (str "mv"))]
+        (->PathProjection [[dim-key uri]] true identity))
+      (->PathProjection nil true identity))))
 
 (defrecord Dataset [uri title description dimensions measures])
 
@@ -226,7 +229,7 @@
 (defn get-dataset-measure-by-uri [{:keys [measures] :as dataset} uri]
   (util/find-first #(= uri (:uri %)) measures))
 
-(defn dataset-result-projection [dataset]
+(defn dataset-result-projection [dataset bindings]
   (into {} (map (fn [{:keys [field-name] :as ft}]
-                  [field-name (get-result-projection ft)])
+                  [field-name (get-result-projection ft bindings)])
                 (dataset-dimension-measures dataset))))

